#7 feature(lab7): Реализовал 7 лабораторную работу

inittose opened the pull request:
- Добавил статический класс DataTools, который работает с данными товаров
- Реализовал функции с делегатом
- На вкладке ItemsTab реализовал логику поиска и сортировки товаров

inittose committed:
feature(lab4): Реализовал класс PriorityOrder

- Создал перечисление OrderTime

inittose committed:
feature(lab4): Добавил верстку PriorityOrdersTab

inittose committed:
feature(lab4): Реализовал логику работы Priority

- Добавил Priority для Customer
- Добавил настройку для сериализатора TypeNameHandling

inittose committed:
docs(lab4): Обновил UML-диаграмму

- Добавил пометку для Priority Order

inittose committed:
style(lab4): Добавил комментарии

inittose committed:
docs(lab4): Обновил UML-диаграмму в README

inittose committed:
feature(lab5): Добавил классы скидок

- Добавил новые методы валидации полей
- Добавил расчет общей стоимости товаров с одной категорией

inittose committed:
feature(lab5): Создал интерфейс IDiscount

- Рассортировал классы из Model по папкам

inittose committed:
fix(lab5): Поправил зависимости (using) проекта

inittose committed:
feature(lab5): Реализовал логику добавления скидки

- Также Реализовал всплывающее окно AddDiscountPopUp

inittose committed:
feature(lab5): Реализовал логику скидок

- Скидки применяются в CartsTab
- Скидки сериализуются

inittose committed:
docs(lab5): Обновил UML-диаграмму

- Добавил отсутствующие комментарии

inittose committed:
style(lab4): Расставил пустые строки

- Назвал все свойства по шаблону
- Реализовал логику удаления данных при ошибочной десериализации

inittose committed:
fix(lab5): Класс IDiscount вынес в папку Model

inittose committed:
fix(lab5): Добавил правки из feature/lab4

- По средствам rebase

inittose committed:
feature(lab6): Реализовал 6 лабораторную работу

- Реализовал стандартные интерфейсы ICloneable, IEquatable<>, IComparable<>
- Обновил UML-диаграмму

inittose committed:
feature(lab7): Создал сервисный класс DataTools

- Добавил делигат CompareProperties для фильтрации списка товаров

inittose committed:
feature(lab7): Изменил логику вкладки ItemsTab

- Вкладка теперь работает не по SelectedIndex, а по SelectedItem
- Добавил функцию поиска товаров по названию на вкладке

inittose committed:
feature(lab7): Реализовал логику сортировки товаров

inittose committed:
docs(lab7): Обновил UML-диаграмму

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Address.cs:
```diff
@@ -188,5 +189,47 @@ public Address()
+        public object Clone()
+        {
+            return new Address(
+                this.Index, 
```
Убрать this

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Address.cs:
```diff
@@ -188,5 +189,47 @@ public Address()
+                return true;
+            }
+
+            var result = this.Index == other.Index;
```
убрать this

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Cart.cs:
```diff
@@ -23,5 +25,21 @@ public float Amount
+        {
+            var cart = new Cart();
+
+            foreach(var item in this.Items)
```
убрать this. Исправь во всем проекте.

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Address.cs:
```diff
@@ -1,11 +1,12 @@
+    public class Address : ICloneable, IEquatable<Address>
     {
         /// <summary>
         /// Разряд почтового индекса.
```
Прощу прощения, что раньше не обратил внимания, константы именуются с использованием стиля Паскаль. Исправить во всем проекте :)

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PercentDiscount.cs:
```diff
@@ -0,0 +1,146 @@
+            get => _discount;
+            private set
+            {
+                ValueValidator.AssertIntOnLimits(value, 1, 10, nameof(Discount));
```
Магические числа

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PercentDiscount.cs:
```diff
@@ -0,0 +1,146 @@
+        {
+            var amount = ItemsTool.GetAmountOnCategory(items, Category);
+            SpendingPerCategory += amount;
+            var percentage = (int)(SpendingPerCategory / 1000) + 1;
```
Все магические числа исправить

Nikita-Isaichenko replied:
а тут что за числа, ты так и не исправил

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PointsDiscount.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            var discount = Calculate(items);
+            Points -= (int)discount;
+            return discount;
```
добавить перед return строку пустую

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -155,24 +158,70 @@ public Item(string name, string info, float cost, Category category)
+        public object Clone()
         {
-            _id = id;
+            var item = new Item(this.Id);
```
this

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -155,24 +158,70 @@ public Item(string name, string info, float cost, Category category)
+            item.Info = this.Info;
+            item.Cost = this.Cost;
+            item.Category = this.Category;
+            return item;
```
перед return строку. Везде

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -155,24 +158,70 @@ public Item(string name, string info, float cost, Category category)
+        /// </returns>
+        public int CompareTo(Item other)
+        {
+            if (this.Cost == other.Cost)
```
this

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Orders/Order.cs:
```diff
@@ -63,6 +69,17 @@ public float Amount
+        /// <summary>
+        /// Возращает конечную стоимость заказа.
+        /// </summary>
+        public double Total
```
Тут конечную стоимость ты хранишь в double, в свойстве Amount - в float, привести к одному типу, double. Проверь везде, где хранишь информацию о стоимости и измени на double

CephalonGL replied:
Для хранения валюты нужно использовать decimal

float, double и decimal являются представлениями чисел с плавающей запятой. Из-за особенности представления чисел с плавающей запятой в памяти компьютера, при выполнении арифметических операций над ними, возникает погрешность. Это недопустимо при расчёте денежных средств. Для решения этой проблемы существует специальный тип данных - decimal

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Orders/Order.cs:
```diff
@@ -80,23 +98,46 @@ public Order()
+                return true;
+            }
+
+            return this.Id == other.Id;
```
this

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Orders/PriorityOrder.cs:
```diff
@@ -0,0 +1,38 @@
+        /// <summary>
+        /// Возвращает и задает дату доставки.
+        /// </summary>
+        public DateTime Date { get; set; }
```
Лучше, наверное, DeliveryDate и ниже DeliveryTime, чтобы было сразу понятно, что означает эта переменная

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Orders/PriorityOrder.cs:
```diff
@@ -0,0 +1,38 @@
+            List<Item> items,
+            double discountAmount) : base(status, address, items, discountAmount)
+        {
+
```
Убрать лишнюю строку

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/CustomerFactory.cs:
```diff
@@ -25,19 +25,18 @@ public static class CustomerFactory
         private const int MAX_POST_INDEX = 999999;
```
Именование констант

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/DataTools.cs:
```diff
@@ -0,0 +1,71 @@
+namespace ObjectOrientedPractics.Services
+{
+    /// <summary>
+    /// Фильтрует и сортирует данные.
```
Я бы в комментарии написал: "Хранит логику по обработке данных". Сейчас название класса слишком обобщенное по сравнению с тем, что указано в комментарии

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CartsTab.cs:
```diff
@@ -9,10 +11,30 @@ namespace ObjectOrientedPractics.View.Tabs
+            set
+            {
+                var isVisible = value >= 0;
+                AmountLabel.Visible = isVisible;
```
Установку свойств видимости можно вынести в отдельный метод в этом же классе, чтобы убрать лишний код из сеттера

Nikita-Isaichenko commented on ObjectOrientedPractics/ObjectOrientedPractics/View/Tabs/CartsTab.cs:
```diff
@@ -24,6 +46,8 @@ public partial class CartsTab : UserControl
         /// </summary>
         public List<Customer> Customers { get; set; }
 
+        public double DiscountAmount { get; set; }
```
Комментарий

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Address.cs:
```diff
@@ -188,5 +189,47 @@ public Address()
+        /// </summary>
+        /// <param name="other">Объект класса <see cref="Address"/>.</param>
+        /// <returns>Возвращает булевое значение, равны ли объекты.</returns>
+        public bool Equals(Address other)
```
Реализация метода Equals должна использовать сравнение хеш-кодов двух Address. Метод расчёта хеш-кода уже объявлен в интерфейсе IEquatable

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PointsDiscount.cs:
```diff
@@ -0,0 +1,131 @@
+        {
+            var amount = ItemsTool.GetAmount(items);
+
+            if (Points / amount > 0.3)
```
Убрать магические числа

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PointsDiscount.cs:
```diff
@@ -0,0 +1,131 @@
+        public void Update(List<Item> items)
+        {
+            var amount = ItemsTool.GetAmount(items);
+            Points += (int)Math.Ceiling(amount * 0.1);
```
Магическое число

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Discounts/PointsDiscount.cs:
```diff
@@ -0,0 +1,131 @@
+
+        /// <summary>
+        /// Применяет накопительные баллы на скидку, доступную для списка товаров.
+        /// Скидка товаров не может быть больше 30% от общей суммы товаров.
```
Избегай ссылок на численные параметры в комментариях, потому что если в процессе разработки требования к ПО поменяются и это значение изменится, то это значение в комментарии тоже придётся менять. К тому же, про это место можно забыть и не исправить после изменения значения в программе и из-за этого документация станет недостоверной.

Магическое число вынеси в константу и в комментарии сделай ссылку на созданное поле/свойство константы через see cref

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Enums/OrderTime.cs:
```diff
@@ -0,0 +1,15 @@
+    /// </summary>
+    public enum OrderTime
+    {
+        FromNineToElevenAM,
```
Добавить XML-комментарий к каждому элементу перечисления. Исправить во всём решении

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Model/Item.cs:
```diff
@@ -155,24 +158,70 @@ public Item(string name, string info, float cost, Category category)
+        /// </summary>
+        /// <param name="other">Объект класса <see cref="Item"/>.</param>
+        /// <returns>Возвращает булевое значение, равны ли объекты.</returns>
+        public bool Equals(Item other)
```
Сравнение по хеш-коду. Исправить во всём решении

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/Services/DataTools.cs:
```diff
@@ -0,0 +1,71 @@
+        {
+            var sortedItems = new List<Item>();
+
+            foreach (var item in items)
```
Списки поддерживают клонирование. Использовать встроенный метод Clone() вместо самостоятельного копирования

CephalonGL commented on ObjectOrientedPractics/ObjectOrientedPractics/View/MainForm.cs:
```diff
@@ -49,6 +49,9 @@ private void MainTabControl_SelectedIndexChanged(object sender, System.EventArgs
             switch (MainTabControl.SelectedIndex)
             {
+                case 1:
+                    CustomersTab.UpdateDiscountsListBox();
```
Обернуть содержимое кейсов в фигурные скобки

inittose committed:
fix(lab7): Правки к 7 лабе

- Убрал this
- Добавил блоки к case
- Исправил магические числа
- Исправил именование констант
- Добавил комментарии к элементам Enumeration
- Обновил UML-диграмму, перешел с drawio на EA

CephalonGL commented on doc/Practics7.png:
См. замечания в дс

inittose committed:
fix(lab7): Небольшие правки

- Разделил стрелки в UML-диаграмме
- Удалил старые диаграммы
- Исправил замечания в коде
